### PR TITLE
Introduce `ruff_index` crate

### PR DESCRIPTION
**Pull request by MichaReiser**
We have multiple places where we use a custom `Id` to index into a Vector. I now plan to add one new index and thought it might be good to reduce the necessary boilerplate. 

This PR formalizes this approach by introducing new `IndexVec` and `IndexSlice` new type wrappers and thew `newtype_index` macro that allows you to define your own index. 

This work is heavily inspired by [rustc `index` crate](https://github.com/rust-lang/rust/blob/master/compiler/rustc_index/src/lib.rs) and thanks to @boshen for making me aware of it :)

### CODE REVIEW

**Review by charliermarsh — APPROVED** [review]

**Comment by Boshen** [comment]
Nice! Let me quickly steal some of the code back 😃

**Comment by MichaReiser** [comment]
> Edit: I forgot to search for a crate, and there is one! [crates.io/crates/index_vec](https://crates.io/crates/index_vec)

Oh nice. I think I still go with our implementation so that we can customize the API (e.g. I intentionally omitted all removal APIs).

File: crates/ruff_python_semantic/src/binding.rs [hunk 1]
```diff
@@ -130,70 +130,46 @@ impl<'a> Binding<'a> {
...
-pub struct BindingId(u32);
-
-impl TryFrom<usize> for BindingId {
-    type Error = TryFromIntError;
-
-    fn try_from(value: usize) -> Result<Self, Self::Error> {
-        Ok(Self(u32::try_from(value)?))
-    }
-}
+#[newtype_index]
```
[the remark this answers is not in the document]
Comment (MichaReiser) [reply crates/ruff_python_semantic/src/binding.rs:133]: No, because of the reason you're stating.

**Comment by charliermarsh** [comment]
I'd like to use this for references, so merge at your leisure :)

**Pull request merged**